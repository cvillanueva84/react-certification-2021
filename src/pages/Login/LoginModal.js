import React, { useState } from 'react';
import ReactDOM from 'react-dom';
import { useHistory } from 'react-router';
import { useAuth } from '../../providers/Auth';
import { CloseBtn, ErrorDiv, ErrorSpan, FormBtns, FormInput, FormStrong, LoginBtn, LoginForm, LoginFormGroup, LoginH1, LoginSection, OverlayDiv } from './LoginModal.styled';

function LoginModal({ open, onClose }) {
  const [userInfo, setUserInfo] = useState({
    username: '',
    password: '',
  });
  const [error, setError] = useState('');
  const { login } = useAuth();
  const history = useHistory();
  const { username, password } = userInfo;

  const authenticate = async (e) => {
    e.preventDefault();
    try {
      setError(null);
      await login(userInfo);
      onClose();
      history.push('/');
    } catch (error) {
      setError(error);
      setUserInfo({
        username: '',
        password: '',
      });
    }
  };

  const handleUserInfo = (e) => {
    setUserInfo({
      ...userInfo,
      [e.target.name]: e.target.value,
    });
  };

  return ReactDOM.createPortal(
    <>
      {open && (
        <>
          <OverlayDiv/>
          <LoginSection>
            <LoginH1>Welcome back you little astronaut!</LoginH1>
            <LoginForm onSubmit={authenticate}>
              <LoginFormGroup >
                <label htmlFor="username">
                  <FormStrong>username </FormStrong>
                  <FormInput
                    required
                    type="text"
                    id="username"
                    name="username"
                    value={username}
                    onChange={handleUserInfo}
                    autoComplete="on"
                  />
                </label>
              </LoginFormGroup>
              <LoginFormGroup >
                <label htmlFor="password">
                  <FormStrong>password </FormStrong>
                  <FormInput
                    required
                    type="password"
                    id="password"
                    name="password"
                    value={password}
                    onChange={handleUserInfo}
                    autoComplete="off"
                  />
                </label>
              </LoginFormGroup>
              {error && (
                <ErrorDiv>
                  <ErrorSpan>Incorrect credentials you dummy!</ErrorSpan>
                </ErrorDiv>
              )}
              <FormBtns >
                <LoginBtn type="submit">
                  Login
                  <i className="fas fa-thumbs-up" style={{ marginLeft: '1rem' }} />
                </LoginBtn>
                <CloseBtn type="button" onClick={() => onClose()}>
                  Close
                  <i className="fas fa-window-close" style={{ marginLeft: '1rem' }} />
                </CloseBtn>
              </FormBtns>
            </LoginForm>
          </LoginSection>
        </>
      )}
    </>,
    document.getElementById('portal')
  );
}

export default LoginModal;
