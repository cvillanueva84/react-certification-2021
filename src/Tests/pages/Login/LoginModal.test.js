import { mount } from 'enzyme/build';
import React from 'react';
import LoginModal from '../../../pages/Login/LoginModal';
import AuthProvider from '../../../providers/Auth';


describe('Tests of the Login Modal component', () => {

    let wrapper; 
     // add a div with #portal id to the global body
    const portalRoot = global.document.createElement('div');
    portalRoot.setAttribute('id', 'portal');
    const body = global.document.querySelector('body');
    body.appendChild(portalRoot);
    afterEach(() => {
        wrapper.unmount();
    });

    const login = jest.fn()
    const onClose = jest.fn()
    wrapper = mount(
        <AuthProvider
            value={{
                login
            }}
        >
            < LoginModal
                open={false}
                onClose={onClose}
            />
        </AuthProvider>
    )    

    test('Should render the component correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })
    
    
})
