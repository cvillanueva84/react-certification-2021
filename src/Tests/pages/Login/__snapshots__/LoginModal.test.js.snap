// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of the Login Modal component Should render the component correctly 1`] = `
<Portal
  containerInfo={
    <div
      id="portal"
    />
  }
/>
`;
